File: internal/domain/markdown.go
package domain

import (
	"fmt"
	"strings"
)

func RenderMarkdown(jstDate, audience string, clusters []Cluster, enriched []EnrichedCluster) string {
	em := map[string]EnrichedCluster{}
	for _, e := range enriched {
		em[e.ClusterID] = e
	}
	lines := []string{
		fmt.Sprintf("*今日のエンジニアトレンド（%s JST）*", jstDate),
		fmt.Sprintf("*対象読者:* %s", audience),
	}
	if len(clusters) == 0 {
		lines = append(lines, "(該当なし)")
		return strings.Join(lines, "\n\n")
	}
	for i, c := range clusters {
		e := em[c.ID]
		summary := e.SummaryJA
		if summary == "" {
			summary = "要約生成に失敗したため、元リンクを確認してください。"
		}
		reason := e.ReasonToRead
		if reason == "" {
			reason = "対象読者の実務に関連する可能性が高いため。"
		}
		url := c.CanonicalURL
		if url == "" && len(c.Items) > 0 {
			url = c.Items[0].URL
		}
		lines = append(lines,
			fmt.Sprintf("%d. *%s*\nURL: %s\n要約: %s\nおすすめ理由: %s", i+1, c.Title, url, summary, reason),
		)
	}
	return strings.Join(lines, "\n\n")
}

File: src/domain/markdown.ts
import { Cluster, EnrichedCluster } from "./model.js";

function clusterLine(cluster: Cluster, enriched?: EnrichedCluster): string {
  const links = cluster.items
    .slice(0, 2)
    .map((i) => `<${i.url}|${i.source}>`)
    .join(" / ");

  const summary = enriched?.summaryJa ?? "要約生成に失敗したため、元リンクを確認してください。";
  const tags = (enriched?.tags ?? ["Trend"]).slice(0, 3).map((t) => `#${t}`).join(" ");
  const reason = enriched?.reasonToRead ?? `増分スコア:+${cluster.deltaScore} / コメント:+${cluster.deltaComments}`;

  return [
    `• *${cluster.title}*`,
    `  ${summary}`,
    `  ${tags}`,
    `  理由: ${reason}`,
    `  link: ${links}`
  ].join("\n");
}

function section(title: string, clusters: Cluster[], enrichedMap: Map<string, EnrichedCluster>): string {
  const lines = clusters.map((c) => clusterLine(c, enrichedMap.get(c.id))).join("\n\n");
  return `*${title}*\n${lines || "(なし)"}`;
}

export function renderMarkdown(input: {
  jstDate: string;
  topTopics: Cluster[];
  rising: Cluster[];
  deepDiscussion: Cluster[];
  enriched: EnrichedCluster[];
}): string {
  const enrichedMap = new Map(input.enriched.map((x) => [x.clusterId, x]));
  return [
    `*今日のエンジニアトレンド（${input.jstDate} JST）*`,
    section("🧩 注目トピック Top 7", input.topTopics, enrichedMap),
    section("🔥 急上昇 Top 5", input.rising, enrichedMap),
    section("🧵 議論が深い Top 5", input.deepDiscussion, enrichedMap)
  ].join("\n\n");
}
